collagraph/cgx/cgx.py

```diff
@@ -490,13 +521,27 @@ def visit_Name(self, node):
         )
 
 
+class ImportsCollector(ast.NodeVisitor):
+    def __init__(self):
+        self.names = set()
+
+    def visit_ImportFrom(self, node):
+        for alias in node.names:
+            self.names.add(alias.asname or alias.name)
+
+    def visit_Import(self, node):
+        for alias in node.names:
+            self.names.add(alias.asname or alias.name)
```

Comment:
It is possible that state or props contain the same names and that the user intends to shadow an import and gets some surprising errors when their code resolves to an import.

Reply:
Yes, you are right. I guess that in order to solve that, we should either disable the syntactic sugar (or at least constrain to not having to type self?) or have a way to define props/state so that we don't need the `lookup` function.

Currently, if users do get a name clash, then they can resolve it by being more specific about where the variable comes from, for instance: `state['foo']` instead of just `foo`.

Biggest problem I guess is that this shadowing can fly under the radar.

What do you think?

Reply:
One idea is that you could pass or embed the collected list of names in the generated code so _lookup can check and raise/warn name collisions.

Reply:
But ideally you just let Python handle resolution. Isn't that what _lookup does?

Reply:
> It does leave the situation where the shadowing is intentional and the user _wants_ to refer to the imported name. How to disable the warning then? :)

Well, the user can always alias the imports to something else, right? 🤷‍♂️ I think embracing the syntactic sugar means that we should just warn about shadowing.

Should we have a 'development/prod' switch for this? As we might want to disable these kinds of runtime checks for production builds?

Reply:
If you're using python's `warnings` module, that already gives users control to disable/enable specific warnings, in a sense. But, you could provide such a setting anyway.

Reply:
> One idea is that you could pass or embed the collected list of names in the generated code so _lookup can check and raise/warn name collisions.

This has a bit of a bootstrapping problem: in order to not get any 'imported but unused' flake8 warnings, they can't be wrapped with the `_lookup` method. Which makes it impossible to check them in the `_lookup` function.

But, I can add a statement to the body of the generated `render` function that checks that the imported names are not used within the state/props or is an attribute on `self`.